Update Firmware flashing page

ctzsnooze opened the pull request:
Adds some notes and details to the firmware flashing page:
- explains a simple / basic flash process with an example screen
- includes RACE PRO option and link
- fixes incorrect note about Dashboard functionality
- many small text / editorial / structural changes

Adds a document under developer/defines.md which lists all the current defines, excluding those on commented-out lines and in the `mk` or `config` directories, using this grep:
`grep --exclude-dir=config -Irh '^[^/].*' src/main/ | grep -wo 'USE_[0-9A-Z_]\+' | sort | uniq`

To search for defines that are commented out, use 
`grep --exclude-dir=config -Irhwo 'USE_[0-9A-Z_]\+' src/main/ | sort | uniq`

To include defines that are commented out, including those in your local config directory, use 
`grep -Irhwo 'USE_[0-9A-Z_]\+' src/main/ | sort | uniq`

haslinghuis commented on docs/wiki/configurator/firmware-flasher-tab.md:
```diff
@@ -80,86 +131,118 @@ Select the receiver wire protocol used. The most common are:
 :::note
 
-If using a local configuration file use the same button again to load a local hex file.
+If, when flashing older MPU-generic firmware, and if you have a local configuration file, load it first, then use the same button again to load the local hex file.
```
```suggestion
If, when flashing older MCU-generic firmware, and if you have a local configuration file, load it first, then use the same button again to load the local hex file.
```

ctzsnooze replied:
Done, thanks

haslinghuis commented on docs/wiki/configurator/firmware-flasher-tab.md:
```diff
@@ -4,11 +4,13 @@ sidebar_position: 21
-Here you update your firmware and select the features you require for you build.
+This tab is used to update the flight controller's Firmware.
+
+Since Betaflight 4.4, firmware updating is by an online build process. The user selects their flight controller, the code version to flash, and enables the features they want, then a custom firmware will be built online, downloaded, and installed onto the flight controller.
```
```suggestion
Since Betaflight 4.4, updating the firmware is done via an online building process. The user selects their flight controller, the code version to flash, and enables the features they want, then a custom firmware will be built online, downloaded, and installed onto the flight controller.
```

ctzsnooze replied:
Done, thanks!

haslinghuis replied:
Either my browser is not updating or it has not been committed :)

ctzsnooze commented:
I note that 'FLASH' and 'FRSKY_OSD' are not options anymore.

haslinghuis commented on docs/wiki/configurator/firmware-flasher-tab.md:
```diff
@@ -155,8 +157,8 @@ For CRSF, ELRS, FPORT or GHST protocols, this is included by default to simplify
 | ESC Serial (SK) Inc. 4way | Enable SimonK and ESC Serial support for flashing via 4way interface                                          |
-| Flash Storage             | Enable Blackbox Flash Storage                                                                                 |
-| FrSky OSD                 | Enable FrSky OSD support                                                                                      |
+| Flash Storage             | Auto-included in 4.5, enables Blackbox Flash Storage before 4.5                                               |
+| FrSky OSD                 | Auto-included in 4.5, enables FrSky OSD support before 4.5                                                    |
```
-  `Flash Storage` aka `USE_FLASH` - is not automatically included - only if provided by the config.
- `FRSKYOSD` - needs to be included using Other Options

ctzsnooze replied:
What I'm saying is the option `Flash Storage` no longer shows up under `Other Options`.  It simply doesn't exist.  So I propose:
`Flash Storage - 4.4 and earlier only.  Enables Blackbox Flash Storage support.  In 4.5, is normally auto-included via the FC config file.  To manually add Flash to a 4.5 build, enter `FLASH` in Custom Defines.`
Likewise for FrSkyOSD:
`4.4 and earlier only. For 4.5, enter `FRSKYOSD` in `Custom Defines``

ctzsnooze commented on docs/wiki/configurator/firmware-flasher-tab.md:
```diff
@@ -180,6 +182,8 @@ It allows the user to enter the coded 'name' for a 'hardware define', forcing th
 
 For example, the local build options `-DUSE_RANGEFINDER -DUSE_ACCGYRO_LSM6DSO` can be included as Custom Defines when making a cloud build with `RANGEFINDER ACCGYRO_LSM6DSO`.
 
+This is a full listing of [Custom Defines](https://hackmd.io/@nerdCopter/H11rvS8Lh), current as of early 2025. Most hardware defines are already included in the board config. Some relate to archaic code that may not function properly and is not currently included in normal builds; test these at your own risk.
```
@nerdCopter I updated the image.
Maybe you could put the list into betaflight.com/docs/development?
That would be awesome.
If it was a table, we could add info about each option (whether it's actively updated, safe to use, what it does, deprecated, etc.
Then I can link to that page?

haslinghuis replied:
Should use this link: https://betaflight.com/docs/development/API/Cloud-Build-API#custom-defines

haslinghuis replied:
Will go through updating custom and maybe other defines on that page if needed asap.

nerdCopter replied:
@ctzsnooze , for now remove this paragraph and link to aforementioned doc instead.  
paragraph wrongly said `2025` anyway.

nerdCopter replied:
https://github.com/betaflight/betaflight.com/pull/419

ctzsnooze replied:
updated the link to the defines.md file  in the cloud-build-api document

haslinghuis commented on docs/wiki/configurator/firmware-flasher-tab.md:
```diff
@@ -150,27 +150,27 @@ For CRSF, ELRS, FPORT or GHST protocols, this is included by default to simplify
+| EMFAT (AutoRun, Icon)     | Enable FAT emulation and icon for onboard flash or MSC                                                                                                                                 |
+| ESC Serial (SK) Inc. 4way | Enable SimonK and ESC Serial support for flashing via 4way interface                                                                                                                   |
+| Flash Storage             | 4.4 and earlier only. Enables Blackbox Flash Storage support. In 4.5, is auto-included via the FC config file. To manually add Flash to a 4.5 build, enter `FLASH` in `Custom Defines` |
+| FrSky OSD                 | 4.4 and earlier only. For 4.5, enter `FRSKYOSD` in `Custom Defines`                                                                                                                    |
```
FrSky OSD is available in Other Options.

ctzsnooze replied:
Thank you - for some odd reason I thought it wasn't.  Fixed.

haslinghuis commented on docs/wiki/configurator/firmware-flasher-tab.md:
```diff
@@ -80,86 +133,122 @@ Select the receiver wire protocol used. The most common are:
-### Other Options
+For example, the local build options `-DUSE_RANGEFINDER -DUSE_ACCGYRO_LSM6DSO` can be included as Custom Defines when making a cloud build with `RANGEFINDER ACCGYRO_LSM6DSO`.
+
+This is a full listing of [Custom Defines](https://hackmd.io/@nerdCopter/H11rvS8Lh), current as of early 2025. Most hardware defines are already included in the board config. Some relate to archaic code that may not function properly and is not currently included in normal builds; test these at your own risk.
```
```suggestion
This is a full listing of [Custom Defines](https://betaflight.com/docs/development/API/Cloud-Build-API#custom-defines). Most hardware defines are already included in the board config. Some relate to archaic code that may not function properly and is not currently included in normal builds; test these at your own risk.
```

ctzsnooze commented:
Quite a big update in the last commit:
- Added, and linked to, a `Defines` file in `development` with the grep code to build it.  Thanks @nerdcopter.  The file is user-focused and includes a simple explanation of the basic usage and purpose of defines.
- Changed the ordering of the configurator tabs so that Firmware Flasher is at the top, and removed the junk `modes` tab document (modes tab info is via `auxiliary-tab.md`)
- Various editorial fixes and tidy-ups.

At the time I did this I didn't notice @nerdCopter's list of defines being added to the cloud build API page.

Defines are used more widely than just for cloud build purposes.  I think that it's best to focus the cloud build page on the defines made visible to the user in the cloud build process.  It's OK to have a complete defines list there as well, but since the full list has relevance to non-cloud build users, perhaps having the complete list in a more generic page about Defines may work a bit better?

nerdCopter commented on docs/development/Defines.md:
```diff
@@ -0,0 +1,531 @@
+
+Additional defines may be added:
+
+- to a cloud build, by entering the name of the define, \_without the USE\_\_ prefix, separated by spaces, in the `Custom Defines` list, e.g. `ANA EXTI`.
```
fix two of the underscored to italics, not rendering.
```suggestion
- to a cloud build, by entering the name of the define, _without the USE\__ prefix, separated by spaces, in the `Custom Defines` list, e.g. `ANA EXTI`.
```

ctzsnooze replied:
Thanks I'll just not italicise that area for simplicity

ctzsnooze commented:
Updated the list of defines, using:
`grep --exclude-dir=config -Irh '^[^/].*' src/ | grep -wo 'USE_[0-9A-Z_]\+' | sort | uniq`

As noted above, this includes, for example:
```
USE_HAL_ADC_REGISTER_CALLBACKS
USE_HAL_CEC_REGISTER_CALLBACKS
USE_HAL_COMP_REGISTER_CALLBACKS
USE_HAL_CORDIC_REGISTER_CALLBACKS
USE_HAL_CRYP_REGISTER_CALLBACKS
USE_HAL_DAC_REGISTER_CALLBACKS
USE_HAL_DCMI_REGISTER_CALLBACKS
USE_HAL_DFSDM_REGISTER_CALLBACKS
USE_HAL_DMA2D_REGISTER_CALLBACKS
USE_HAL_DRIVER
USE_HAL_DSI_REGISTER_CALLBACKS
USE_HAL_DTS_REGISTER_CALLBACKS
USE_HAL_ETH_REGISTER_CALLBACKS
USE_HAL_EXTI_REGISTER_CALLBACKS
USE_HAL_FDCAN_REGISTER_CALLBACKS
USE_HAL_FMAC_REGISTER_CALLBACKS
USE_HAL_GFXMMU_REGISTER_CALLBACKS
USE_HAL_HASH_REGISTER_CALLBACKS
USE_HAL_HCD_REGISTER_CALLBACKS
USE_HAL_HRTIM_REGISTER_CALLBACKS
USE_HAL_I2C_REGISTER_CALLBACKS
USE_HAL_I2S_REGISTER_CALLBACKS
USE_HAL_IRDA_REGISTER_CALLBACKS
USE_HAL_JPEG_REGISTER_CALLBACKS
USE_HAL_LPTIM_REGISTER_CALLBACKS
USE_HAL_LTDC_REGISTER_CALLBACKS
USE_HAL_MDIOS_REGISTER_CALLBACKS
USE_HAL_MMC_REGISTER_CALLBACKS
USE_HAL_NAND_REGISTER_CALLBACKS
USE_HAL_NOR_REGISTER_CALLBACKS
USE_HAL_OPAMP_REGISTER_CALLBACKS
USE_HAL_OSPI_REGISTER_CALLBACKS
USE_HAL_OTFDEC_REGISTER_CALLBACKS
USE_HAL_PCD_REGISTER_CALLBACK
USE_HAL_PCD_REGISTER_CALLBACKS
USE_HAL_QSPI_REGISTER_CALLBACKS
USE_HAL_RNG_REGISTER_CALLBACKS
USE_HAL_RTC_REGISTER_CALLBACKS
USE_HAL_SAI_REGISTER_CALLBACKS
USE_HAL_SDRAM_REGISTER_CALLBACKS
USE_HAL_SD_REGISTER_CALLBACKS
USE_HAL_SMARTCARD_REGISTER_CALLBACKS
USE_HAL_SMBUS_REGISTER_CALLBACKS
USE_HAL_SPDIFRX_REGISTER_CALLBACKS
USE_HAL_SPI_REGISTER_CALLBACKS
USE_HAL_SRAM_REGISTER_CALLBACKS
USE_HAL_SWPMI_REGISTER_CALLBACKS
USE_HAL_TIM_REGISTER_CALLBACKS
USE_HAL_UART_REGISTER_CALLBACKS
USE_HAL_USART_REGISTER_CALLBACKS
USE_HAL_WWDG_REGISTER_CALLBACKS
```
They all start with `USE_`, but only `USE_HAL_DRIVER` a define which we care about from a configuration perspective.  The others just set internal states and are not user-configurable.  I'll make a firmware PR to change 'USE_` to `ENABLE_` or something, then rebuild the list.

nerdCopter commented:
> At the time I did this I didn't notice @nerdCopter's list of defines being added to the cloud build API page.

I'm okay with whatever.  if this new `docs/development/Defines.md` goes in, then the API page need to link to it.

haslinghuis commented:
@ctzsnooze Should change src to src/main otherwise you pick up middleware etc.

ctzsnooze commented:
I think this is ready; from my perspective, please give final reviews / approval if OK.

haslinghuis commented on docs/development/Defines.md:
```diff
@@ -0,0 +1,533 @@
+# Defines
+
+Betaflight uses `#DEFINE` syntax to include software components to support additional features at build time.
```
```suggestion
Betaflight uses `#define` macro syntax to include software components to support additional features at build time.
```

ctzsnooze replied:
Thanks. Updated.

haslinghuis commented on docs/development/Defines.md:
```diff
@@ -0,0 +1,533 @@
+
+Not all the defines in this list are 'safe'; some are archaic and some are not actively maintained.
+Defines on commented-out lines are not included in the list.
+Some defines are for internal register definitions and are not end-user code options.
```
@ctzsnooze suggest leaving out middleware definitions.

ctzsnooze replied:
By prepping only src/main we remove all middleware `USE_` defines.  The only unnecessary ones are those where an internal register definition commences with `USE_`.  Fortunately, not many of them.
Updated the listing now.

haslinghuis commented on docs/development/Defines.md:
```diff
@@ -0,0 +1,533 @@
+USE_CRAFTNAME_MSGS
+USE_CRSF_CMS_TELEMETRY
+USE_CRSF_LINK_STATISTICS
+USE_CRSF_V3
```
```suggestion
USE_CRSF_OFFICIAL_SPEC
USE_CRSF_V3
```

ctzsnooze replied:
ha ha OK

haslinghuis reviewed (approved):
Good work Chris!

nerdCopter commented on docs/development/Defines.md:
```diff
@@ -0,0 +1,533 @@
+
+The user may review the full list of defines in the build by checking the build log after flashing. A link to the log appears in the Flashing tab once flashing is complete. Later, it can be loaded from Configurator's main Setup page, using the `Log` button at bottom right. A summary of the included build options can also be displayed using the nearby `Options` button.
+
+A full list of defines as of May 2024 is provided below. The list was generated using `grep --exclude-dir=config -Irh '^[^/].*' src/main/ | grep -wo 'USE_[0-9A-Z_]\+' | sort | uniq` - thanks @nerdcopter!
```
maybe i'm given too much credit here since @ctzsnooze improved the grepping significantly. 
can leave me out of the doc itself, sionce i'm mentioned in the comments/description.
```suggestion
A full list of defines as of May 2024 is provided below. The list was generated using `grep --exclude-dir=config -Irh '^[^/].*' src/main/ | grep -wo 'USE_[0-9A-Z_]\+' | sort | uniq`
```
